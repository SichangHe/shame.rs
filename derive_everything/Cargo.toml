[package]
name = "derive_everything"
version = "0.1.0"
edition = "2021"
description = "#[derive(Copy, Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)] everything if possible."
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lib]
proc-macro = true

[dev-dependencies]
trybuild = { version = "1", features = ["diff"] }

[dependencies]
impls.workspace = true
quote = "1"
syn = { version = "2", features = ["full"] }
