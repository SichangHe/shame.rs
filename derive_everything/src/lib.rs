use proc_macro::TokenStream;
use quote::quote;
use syn::{parse::Nothing, parse_macro_input, Item};

#[proc_macro_attribute]
pub fn derive_everything(args: TokenStream, input: TokenStream) -> TokenStream {
    let _ = parse_macro_input!(args as Nothing);
    let item = parse_macro_input!(input as Item);

    quote! {
        #[derive(Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
        #item
    }
    .into()
}

#[proc_macro_attribute]
pub fn derive_enum_everything(args: TokenStream, input: TokenStream) -> TokenStream {
    let _ = parse_macro_input!(args as Nothing);
    let item = parse_macro_input!(input as Item);

    quote! {
        #[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
        #item
    }
    .into()
}

#[proc_macro_attribute]
pub fn derive_float_everything(args: TokenStream, input: TokenStream) -> TokenStream {
    let _ = parse_macro_input!(args as Nothing);
    let item = parse_macro_input!(input as Item);

    quote! {
        #[derive(Clone, Debug, Default, PartialEq, PartialOrd)]
        #item
    }
    .into()
}

#[proc_macro_attribute]
pub fn derive_float_enum_everything(args: TokenStream, input: TokenStream) -> TokenStream {
    let _ = parse_macro_input!(args as Nothing);
    let item = parse_macro_input!(input as Item);

    quote! {
        #[derive(Clone, Debug, PartialEq, PartialOrd)]
        #item
    }
    .into()
}
