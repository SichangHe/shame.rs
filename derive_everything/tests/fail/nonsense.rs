use derive_everything::derive_everything;

#[derive_everything]
fn noop() {}

#[derive_everything]
trait Trait {}

fn main() {}
