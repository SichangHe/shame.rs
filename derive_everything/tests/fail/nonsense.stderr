error[E0774]: `derive` may only be applied to `struct`s, `enum`s and `union`s
 --> tests/fail/nonsense.rs:3:1
  |
3 | #[derive_everything]
  | ^^^^^^^^^^^^^^^^^^^^ not applicable here
4 | fn noop() {}
  | ------------ not a `struct`, `enum` or `union`
  |
  = note: this error originates in the attribute macro `derive_everything` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0774]: `derive` may only be applied to `struct`s, `enum`s and `union`s
 --> tests/fail/nonsense.rs:6:1
  |
6 | #[derive_everything]
  | ^^^^^^^^^^^^^^^^^^^^ not applicable here
7 | trait Trait {}
  | -------------- not a `struct`, `enum` or `union`
  |
  = note: this error originates in the attribute macro `derive_everything` (in Nightly builds, run with -Z macro-backtrace for more info)
